File: 0017-Letter-Combinations-of-a-Phone-Number/0017.py
class Solution:
    def __init__(self):
        self.letterMap = [
            ' ',
            '',
            'abc',
            'def',
            'ghi',
            'jkl',
            'mno',
            'pqrs',
            'tuv',
            'wxyz'
        ]
    def findCombination(self, digits, index, s, res):
        if index == len(digits):
            res.append(s)
            return 

        char = digits[index]
        letters = self.letterMap[ord(char) - ord('0')]
        for letter in letters:
            self.findCombination(digits, index + 1, s + letter, res)

    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        result = list()
        if not digits:
            return result
            
        self.findCombination(digits, 0, "", result)
        return result

File: src/0017-Letter-Combinations-of-a-Phone-Number/0017.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    vector<string> letterCombinations(string digits) 
    {
        vector<string> result;
        if (digits.empty()) return result;
        findCombination(digits, 0, string(""), result);
        return result;
    }
private:
    const vector<string> letterMap = {            
                        " ","","abc",
                        "def","ghi","jkl",
                        "mno","pqrs","tuv","wxyz"};

    void findCombination(const string& digits, int index, const string& s, vector<string>& res)
    {
        if (digits.length() == index) 
        {
            res.push_back(s);
            return;
        }
        char ch = digits[index];
        string letters = letterMap[ch - '0'];
        for (auto& letter : letters)
        {
            findCombination(digits, index + 1, s + letter, res);
        }
    }
};